From edf0ac040ad08e386a430c4d53af83670962f614 Mon Sep 17 00:00:00 2001
From: Wei Qin <qinwei61@huawei.com>
Date: Fri, 28 Aug 2026 11:23:12 +0800
Subject: [PATCH 1/1] udma: add log function interfaces.

driver inclusion
category: feature
bugzilla: https://atomgit.com/openeuler/umdk/issues/887
CVE: NA

-----------------------------------------------------------

This patch add log function which could record log
in the container.

Signed-off-by: Wei Qin <qinwei61@huawei.com>
---
 src/urma/hw/udma/udma_u_jetty.c |  2 +-
 src/urma/hw/udma/udma_u_jfc.c   |  8 ++---
 src/urma/hw/udma/udma_u_log.c   | 61 ++++++++++++++++++++++++++++-----
 src/urma/hw/udma/udma_u_log.h   | 40 ++++++++++++++-------
 src/urma/hw/udma/udma_u_ops.c   |  2 ++
 5 files changed, 87 insertions(+), 26 deletions(-)

diff --git a/src/urma/hw/udma/udma_u_jetty.c b/src/urma/hw/udma/udma_u_jetty.c
index 1c979f0..11888b6 100644
--- a/src/urma/hw/udma/udma_u_jetty.c
+++ b/src/urma/hw/udma/udma_u_jetty.c
@@ -724,7 +724,7 @@ urma_status_t udma_u_bind_jetty_ex(urma_jetty_t *jetty,
 	}
 
 	if (udma_jetty->sq.tjetty == udma_tjetty) {
-		UDMA_LOG_INFO("reentry bind JETTY in exp, jetty_id = %u, tjetty_id = %u.\n",
+		UDMA_LOG_WARN("reentry bind JETTY in exp, jetty_id = %u, tjetty_id = %u.\n",
 			      jetty->jetty_id.id, tjetty->id.id);
 		return URMA_SUCCESS;
 	}
diff --git a/src/urma/hw/udma/udma_u_jfc.c b/src/urma/hw/udma/udma_u_jfc.c
index 3682f59..dc71d39 100644
--- a/src/urma/hw/udma/udma_u_jfc.c
+++ b/src/urma/hw/udma/udma_u_jfc.c
@@ -647,25 +647,25 @@ static bool udma_u_update_jfr_idx(struct udma_u_context *udma_ctx,
 		if (udma_ctx->jetty_table[table_id].refcnt) {
 			jetty = (struct udma_u_jetty *)udma_ctx->jetty_table[table_id].jetty_array[jetty_id & mask];
 			if (!jetty) {
-				UDMA_LOG_INFO("JETTY not found, JETTY 0x%x has been destroyed.\n", jetty_id);
+				UDMA_LOG_WARN("JETTY not found, JETTY 0x%x has been destroyed.\n", jetty_id);
 				return true;
 			}
 			cr->user_data = (uintptr_t)&jetty->base;
 			jfr = jetty->jfr;
 		} else {
-			UDMA_LOG_INFO("JFC not polled, JETTY 0x%x has been destroyed.\n", jetty_id);
+			UDMA_LOG_WARN("JFC not polled, JETTY 0x%x has been destroyed.\n", jetty_id);
 			return true;
 		}
 	} else {
 		if (udma_ctx->jfr_table[table_id].refcnt) {
 			jfr = (struct udma_u_jfr *)udma_ctx->jfr_table[table_id].jfr_array[jetty_id & mask];
 			if (!jfr) {
-				UDMA_LOG_INFO("JETTY not found, JETTY 0x%x has been destroyed.\n", jetty_id);
+				UDMA_LOG_WARN("JETTY not found, JETTY 0x%x has been destroyed.\n", jetty_id);
 				return true;
 			}
 			cr->user_data = (uintptr_t)&jfr->base;
 		} else {
-			UDMA_LOG_INFO("JFC not polled, JFR 0x%x has been destroyed.\n", jetty_id);
+			UDMA_LOG_WARN("JFC not polled, JFR 0x%x has been destroyed.\n", jetty_id);
 			return true;
 		}
 	}
diff --git a/src/urma/hw/udma/udma_u_log.c b/src/urma/hw/udma/udma_u_log.c
index d1143f0..1ff41f2 100644
--- a/src/urma/hw/udma/udma_u_log.c
+++ b/src/urma/hw/udma/udma_u_log.c
@@ -16,12 +16,9 @@
 #include "udma_u_log.h"
 
 static struct udma_vlog_level_st g_udma_log_level = {UDMA_VLOG_LEVEL_INFO, PTHREAD_MUTEX_INITIALIZER};
-#define MAX_LOG_LEN 512
-#define UDMA_LOG_TAG "UDMA_LOG_TAG"
-#define UDMA_LOG_ENV_STR "UDMA_LOG_LEVEL"
-#define UDMA_LOG_LEVEL_ENV_MAX_BUF_LEN 32
+static urma_log_cb_t g_udma_log_cb = udma_default_log_func;
 
-static void udma_default_log_func(int level, char *message)
+void udma_default_log_func(int level, char *message)
 {
 	syslog(level, "%s", message);
 }
@@ -81,8 +78,8 @@ int udma_vlog(const char *function, int line, enum udma_vlog_level level, const
 	char logmsg[MAX_LOG_LEN + 1] = {};
 	int ret;
 
-	/* add log head info, "UDMA_LOG_TAG|function|[line]|format" */
-	ret = snprintf(newformat, MAX_LOG_LEN, "%s|%s[%d]|%s", UDMA_LOG_TAG, function, line, format);
+	/* add log head info, "[UDMA_LOG_TAG][function:line] format" */
+	ret = snprintf(newformat, MAX_LOG_LEN, "[%s][%s:%d] %s", UDMA_LOG_TAG, function, line, format);
 	if (ret <= 0 || ret >= (int)sizeof(newformat))
 		return ret;
 
@@ -91,7 +88,7 @@ int udma_vlog(const char *function, int line, enum udma_vlog_level level, const
 		(void)printf("logmsg size exceeds MAX_LOG_LEN size :%d.\n", MAX_LOG_LEN);
 		return ret;
 	}
-	udma_default_log_func((int)level, logmsg);
+	g_udma_log_cb((int)level, logmsg);
 
 	return ret;
 }
@@ -104,3 +101,51 @@ void udma_log(const char *function, int line, enum udma_vlog_level level, const
 	(void)udma_vlog(function, line, level, format, va);
 	va_end(va);
 }
+
+static void udma_u_try_register_ummu_log_backend(void)
+{
+	ummu_log_backend_t backend = {.emit = g_udma_log_cb};
+	int ret;
+
+	if (!udma_u_ummu_log_supported()) {
+		UDMA_LOG_WARN("libummu log backend not supported, skipping.\n");
+		return;
+	}
+
+	ret = ummu_log_register_backend(&backend);
+	if (ret)
+		UDMA_LOG_ERR("Failed to register ummu log backend, ret=%d.\n", ret);
+}
+
+static void udma_u_try_restore_ummu_log_default(void)
+{
+	if (!udma_u_ummu_log_supported()) {
+		UDMA_LOG_WARN("libummu log restore default not supported, skipping.\n");
+		return;
+	}
+	ummu_log_restore_default();
+}
+
+urma_status_t udma_u_register_log_func(urma_log_cb_t func)
+{
+	if (func == NULL) {
+		UDMA_LOG_ERR("Invalid parameter: log callback func is NULL.\n");
+		return URMA_EINVAL;
+	}
+	g_udma_log_cb = func;
+	UDMA_LOG_INFO("libudma registered log successfully.\n");
+
+	udma_u_try_register_ummu_log_backend();
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_unregister_log_func(void)
+{
+	udma_u_try_restore_ummu_log_default();
+
+	UDMA_LOG_INFO("libudma restored to default log output successfully.\n");
+	g_udma_log_cb = udma_default_log_func;
+
+	return URMA_SUCCESS;
+}
diff --git a/src/urma/hw/udma/udma_u_log.h b/src/urma/hw/udma/udma_u_log.h
index 0689069..b27659e 100644
--- a/src/urma/hw/udma/udma_u_log.h
+++ b/src/urma/hw/udma/udma_u_log.h
@@ -12,6 +12,24 @@
 
 #include <pthread.h>
 #include <stdbool.h>
+#include "urma_types.h"
+#include "ummu_api.h"
+
+#define MAX_LOG_LEN 512
+#define UDMA_LOG_TAG "UDMA"
+#define UDMA_LOG_ENV_STR "UDMA_LOG_LEVEL"
+#define UDMA_LOG_LEVEL_ENV_MAX_BUF_LEN 32
+
+#define UDMA_LOG(l, ...)								\
+	do {										\
+		if (!udma_log_drop(UDMA_VLOG_LEVEL_##l))				\
+			udma_log(__func__, __LINE__, UDMA_VLOG_LEVEL_##l, __VA_ARGS__); \
+	} while (0)
+
+#define UDMA_LOG_INFO(...) UDMA_LOG(INFO, __VA_ARGS__)
+#define UDMA_LOG_ERR(...) UDMA_LOG(ERR, __VA_ARGS__)
+#define UDMA_LOG_WARN(...) UDMA_LOG(WARNING, __VA_ARGS__)
+#define UDMA_LOG_DEBUG(...) UDMA_LOG(DEBUG, __VA_ARGS__)
 
 enum udma_vlog_level {
 	UDMA_VLOG_LEVEL_EMERG,
@@ -35,19 +53,15 @@ void udma_getenv_log_level(void);
 void udma_log(const char *function, int line, enum udma_vlog_level level,
 	      const char *format, ...) __attribute__((format(printf, 4, 5)));
 enum udma_vlog_level udma_log_get_level_from_string(const char* level_string);
+urma_status_t udma_u_register_log_func(urma_log_cb_t func);
+urma_status_t udma_u_unregister_log_func(void);
+void udma_default_log_func(int level, char *message);
+__attribute__((weak)) int ummu_log_register_backend(ummu_log_backend_t *backend);
+__attribute__((weak)) void ummu_log_restore_default(void);
 
-#define UDMA_LOG(l, ...)								\
-	do {										\
-		if (!udma_log_drop(UDMA_VLOG_LEVEL_##l))				\
-			udma_log(__func__, __LINE__, UDMA_VLOG_LEVEL_##l, __VA_ARGS__); \
-	} while (0)
-
-#define UDMA_LOG_INFO(...) UDMA_LOG(INFO, __VA_ARGS__)
-
-#define UDMA_LOG_ERR(...) UDMA_LOG(ERR, __VA_ARGS__)
-
-#define UDMA_LOG_WARN(...) UDMA_LOG(WARNING, __VA_ARGS__)
-
-#define UDMA_LOG_DEBUG(...) UDMA_LOG(DEBUG, __VA_ARGS__)
+static inline bool udma_u_ummu_log_supported(void)
+{
+	return (ummu_log_register_backend != NULL && ummu_log_restore_default != NULL);
+}
 
 #endif /* __UDMA_U_LOG_H__ */
diff --git a/src/urma/hw/udma/udma_u_ops.c b/src/urma/hw/udma/udma_u_ops.c
index 0e7a431..964aeca 100644
--- a/src/urma/hw/udma/udma_u_ops.c
+++ b/src/urma/hw/udma/udma_u_ops.c
@@ -380,4 +380,6 @@ const urma_provider_ops_t g_udma_provider_ops = {
 	.query_device = udma_u_query_device,
 	.create_context = udma_u_create_context,
 	.delete_context = udma_u_delete_context,
+	.register_log_func = udma_u_register_log_func,
+	.unregister_log_func = udma_u_unregister_log_func,
 };
-- 
2.28.0.windows.1

